### PR TITLE
Migrate doc from Sphinx to Documenter

### LINKED ISSUE #346: Migrate doc from Sphinx to Documenter
https://github.com/JuliaCollections/DataStructures.jl/issues/345#issuecomment-343831010

What are the pros and cons?

### PR DESCRIPTION
**Pull request by scls19fr**
Convert rst to md using https://gist.github.com/zaiste/77a946bbba73f5c4d33f3106a494e6cd
Add Julia syntax highlighting to md files
Add make.jl for Documenter.jl
Autodeploy doc in .travis.yml
Closes #346

### CODE REVIEW

**Comment by scls19fr** [comment]
Pandoc have been used to convert reStructuredText to Markdown

```
FILES=*.rst
for f in $FILES
do
  filename="${f%.*}"
  echo "Converting $f to $filename.md"
  `pandoc $f -f rst -t markdown -o $filename.md`
done
```

**Comment by scls19fr** [comment]
https://github.com/JuliaCollections/DataStructures.jl/pull/347/files#diff-333026e9bbc6f22aa98e849abacab9d7

doc/src/sorted_containers.md

need some manual improvements

**Comment by scls19fr** [comment]
Definition list in [doc/src/sorted_containers.md](https://github.com/scls19fr/DataStructures.jl/blob/c31f17f0d0dead44ae7766b36fd880dcae666a9d/doc/src/sorted_containers.md) is not rendered correctly.

Any idea how to fix this?

**Comment by kmsquire** [comment]
@scls19fr, thanks for doing this!  I've been going through my messages in order, and finally caught up to this one (so most of my previous comments were out of date).

Not sure I'll have more time today, but will try to take a closer look over the next couple of days.

**Comment by StephenVavasis** [comment]
Approximately 2-3 years ago when I wrote the code sorted containers, I switched all the DataStructures documentation from MD to RST by hand (with Kevin's permission) precisely because I couldn't figure out how to make definition lists in MD, and I needed definition lists for the sorted containers documentation.

**Comment by scls19fr** [comment]
Fashion is an eternal renewal ;-)
I personally prefer (running) doctests than pretty definition lists.
Maybe a workaround could be to use header h... instead of definition lists.
See also https://www.google.fr/search?q=definition+list+markdown

**Comment by kmsquire** [comment]
Okay, had a look. :-)

Regarding `doc/src/sorted_containers.md` formatting: since the definition lists there are used to document the types and functions/methods defined for sorted containers, the best solution would be to move these definitions into the source code as docstrings, and then follow the `Documenter.jl` [instructions for including those docstrings in the documentation](https://juliadocs.github.io/Documenter.jl/stable/man/guide/#Adding-some-docstrings-1).

Would you be up to doing that here?  It should probably be done for all methods/types, but since sorted containers are the most problematic, I think just doing it there for now would be fine. The rest can be moved gradually--e.g., as the package is being split up (cc: @oxinabox).

**Comment by StephenVavasis** [comment]
Just a remark that some of the functions in the sorted container function list apply to more than one of the sorted containers and hence correspond to more than one location in the source code.

**Comment by scls19fr** [comment]
I moved most definitions in `doc/src/sorted_containers.md` to their respective function definitions (duplicating into several files when required).

File: src/sorted_dict.jl [hunk 1]
```diff
@@ -29,23 +29,93 @@ mutable struct SortedDict{K, D, Ord <: Ordering} <: Associative{K,D}
...
+"""
 SortedDict(ps::Pair...) = SortedDict(Forward, ps)
-SortedDict(o::Ordering, ps::Pair...) = SortedDict(o, ps)
 SortedDict{K,D}(ps::Pair...) where {K,D} = SortedDict{K,D,ForwardOrdering}(Forward, ps)
+
+"""
+    SortedDict(o, k1=>v1, k2=>v2, ...)
+
+Construct a `SortedDict` from the given pairs with the specified
+ordering `o`. If `K` and `V` are not specified, the key type and
```
Comment (kmsquire) [inline src/sorted_dict.jl]: `K` and `V` don't appear here.

File: src/sorted_dict.jl [hunk 2]
```diff
@@ -29,23 +29,93 @@ mutable struct SortedDict{K, D, Ord <: Ordering} <: Associative{K,D}
...
+and
+    SortedDict{K,V}(iter, o=Forward)
+
+Construct a `SortedDict` from an arbitrary iterable object of
+`key=>value` pairs. If `K` and `V` are not specified, the key type
+and value type are inferred from the given iterable. The ordering
+object `o` defaults to `Forward`.
+"""
 SortedDict{K,D}(kv) where {K,D} = SortedDict{K,D}(Forward, kv)
 function SortedDict{K,D}(o::Ord, kv) where {K,D,Ord<:Ordering}
```
Comment (kmsquire) [inline src/sorted_dict.jl]: The function signature here doesn't match the one in the comments above.  Might need to rearrange the definitions a little.

File: src/sorted_dict.jl [hunk 3]
```diff
@@ -63,6 +133,30 @@ end
...
+value-type is `Int`.
+
+If `K` and `V` are not specified, the key type and value type are
+inferred from the given dictionary. The ordering object `o` defaults
+to `Forward`.
+
+See below for more information about ordering.
+"""
 SortedDict(kv, o::Ordering=Forward) = SortedDict(o, kv)
 function SortedDict(o::Ordering, kv)
```
Comment (kmsquire) [inline src/sorted_dict.jl]: Same here.

File: src/sorted_set.jl [hunk 4]
```diff
@@ -5,6 +5,19 @@
 mutable struct SortedSet{K, Ord <: Ordering}
     bt::BalancedTree23{K,Void,Ord}
 
+    """
+        SortedSet(iter, o=Forward)
+    and
```
Comment (kmsquire) [inline src/sorted_set.jl]: I personally think that the "and" lines are not needed (here and elsewhere).

**Comment by scls19fr** [comment]
I've just add references to docstrings into sorted_containers.md

about "and" lines... that's difficult to avoid them (at least in this PR)

Some constructors's docstring will probably have to be modified... but I think it's safer to adress this problem in an other PR.

File: src/sorted_dict.jl (hunk 2 above)
Comment (scls19fr) [reply src/sorted_dict.jl]: The problem is is similar to https://github.com/JuliaCollections/DataStructures.jl/issues/340 and should probably be adressed in an other PR

**Comment by kmsquire** [comment]
This looks good--as you suggested, I'm okay with fixing the remaining issues in future pull requests.

However, there are merge conflicts now.  Are you comfortable fixing these?  I can give some command-line help if you need it.

**Comment by scls19fr** [comment]
My git fu is a bit rusty... So some help to deal with this 2 conflicting files could help.
I can squash commit if you prefer. I can only do it tonight or tomorrow.

**Comment by scls19fr** [comment]
Solving this merge conflict wasn't easy for me (and I don't like this kind of task)... but I think it's now ready for merging.
As told in this comment https://github.com/JuliaCollections/DataStructures.jl/pull/347#issuecomment-345554295 some improvement could be done after this PR being merged.

**Comment by scls19fr** [comment]
Current doc is available at ReadTheDocs http://datastructuresjl.readthedocs.io/en/latest

I wonder if it was manually uploaded or automatically uploaded when commits where made.

Documenter.jl provides a nice doc about hosting on github pages https://juliadocs.github.io/Documenter.jl/stable/man/hosting/

I wonder if we should move out of RTD to GH-pages or not...

and who can do it?

If some improvement to this PR are required please tell me.

**Comment by ViralBShah** [comment]
Should we merge this before #348?

**Comment by kmsquire** [comment]
Sorry, been swamped recently, but I plan to look at this tomorrow.

**Comment by scls19fr** [comment]
Any news about this PR? Is there any problems on which I could help? I noticed that some recent PR involve a lot of changes in several files. I would prefer to avoid to have to rebase this PR.

**Comment by kmsquire** [comment]
Again, swamped (and sick right now). Is anyone else available to

1. Review
2. Make the switch to github pages that (I think) this requires?

**Comment by oxinabox** [comment]
I will try and hit this up in the next few days.

What I want to do is branch this branch over to my own repo,
setup github pages on that.
And actually review what it looks like and that the github pages deployment stuff is working right.
(I've only set that up once before I think i mostly followed https://www.juliabloggers.com/finalizing-your-julia-package-documentation-testing-coverage-and-publishing/
but maybe there is better instruction in the Documentor.jl docs now.)

**Comment by oxinabox** [comment]
Sorry, there were reasons, but anyway I am doing my review right  now.
If I can finish it in 1hr then this will be done

**Comment by oxinabox** [comment]
It seems to be working but I get a couple of errors, that I can't quiet workout why (I think you mentioned them before)

```
Documenter: setting up build directory.
Documenter: expanding markdown templates.
Documenter: building cross-references.
 !! No doc found for reference '[`Bool`](@ref)'. [src/sorted_containers.md]
Documenter: running document checks.
 > checking for missing docstrings.
 > running doctests.
 > checking footnote links.
Documenter: populating indices.
Documenter: rendering document.
 !! Invalid local link: unresolved path
    '@ref' in sorted_containers.md
 ?? TRAVIS_REPO_SLUG       = ""
 ??   should match "github.com/oxinabox/DataStructures.jl.git" (kwarg: repo)
 ?? TRAVIS_PULL_REQUEST    = ""
 ??   deploying if equal to "false"
 ?? TRAVIS_OS_NAME         = ""
 ??   deploying if equal to "linux" (kwarg: osname)
 ?? TRAVIS_JULIA_VERSION   = ""
 ??   deploying if equal to "0.6" (kwarg: julia)
 ?? TRAVIS_BRANCH          = ""
 ?? TRAVIS_TAG             = ""
 ??   deploying if branch equal to "master" (kwarg: latest) or tag is set
 ?? git commit SHA         = f38d438
 ?? DOCUMENTER_KEY exists  = false
 ?? should_deploy          = false
Documenter: skipping docs deployment.
  You can set DOCUMENTER_DEBUG to "true" in Travis to see more information.
```

I'm not sure if this is a blocking or not.
I can't see what is causing it, as those lines don't seem to exist in `/docs/src/sorted_container.md`

I am going to push a few changes over to this branch shortly from my staging branch.

**Comment by oxinabox** [comment]
After being merged, this needs the deploy key added to travis and github.
Which I can do, and will do when I merge in ~24 hours.


People should take a look at the staging site: 
http://white.ucc.asn.au/DataStructures.jl/latest

**Comment by scls19fr** [comment]
http://white.ucc.asn.au/DataStructures.jl/latest/disjoint_sets.html

maybe comments (starting with `#...`) should be aligned

Same for http://white.ucc.asn.au/DataStructures.jl/latest/accumulators.html (see `push!`)

On http://white.ucc.asn.au/DataStructures.jl/latest/default_dict.html
‘Notethatinthelastexample,weneedtouseafunctiontocreateeachnew‘Notethatinthelastexample,weneedtouseafunctiontocreateeachnew
looks odd.

**Comment by scls19fr** [comment]
http://white.ucc.asn.au/DataStructures.jl/latest/trie.html

Being previously a big fan of Python and PEP8 I will change

```julia
t=Trie{Int}()
```

to

```julia
t = Trie{Int}()
```

same for others affectations

```julia
t["Rob"]=42
```
->

```julia
t["Rob"] = 42
```

same for

```julia
t["Roger"]=24
```
->
```julia
t["Roger"] = 24
```

I will add a space after coma, and two spaces before `#` and 1 space after `#`

```julia
haskey(t,"Rob") #true
```

->

```julia
haskey(t, "Rob")  # true
```

Same for

```julia
get(t,"Rob",nothing) #42
```
->

```julia
get(t, "Rob", nothing)  # 42
```

but these cosmetic changes could be tackle later if you prefer.

**Comment by oxinabox** [comment]
Do it now, or do it later, I don't mind.

**Comment by oxinabox** [comment]
I have now looked over the output and am satisfied nothing has gone critically wrong.

**Comment by scls19fr** [comment]
Ok so let's publish it now under https://JuliaCollections.github.io/DataStructures.jl/latest/ and fix remaining issues later.

**Comment by oxinabox** [comment]
I've added the hooks so that when I (or someone else) merges this tomorrowish, then the docs  should just build and deploy.

**Comment by oxinabox** [comment]
Ok,  Done 🎉 
Thank you for your work, it was fairly mammoth.
Sorry about how long it took for the code review to happen.

**Comment by oxinabox** [comment]
Yes, I just clicked merge, and 5 minutes later (once travis run) new docs appeared at the new site.
Now I need to workout how to take down the old ones.

**Comment by scls19fr** [comment]
Thanks @oxinabox for merging

**Pull request merged**